src/synchronizeContacts.js

```diff
@@ -1,30 +1,202 @@
+  const resourceName = get(
+    cozyContact,
+    ['cozyMetadata', 'sync', contactAccountId, 'id'],
+    null
+  )
+
+  if (resourceName) {
+    return googleContacts.find(
+      googleContact => googleContact.resourceName === resourceName
+    )
+  }
+
+  return undefined
+}
+
+const findCozyContact = (googleContact, cozyContacts, contactAccountId) => {
+  return cozyContacts.find(
+    contact =>
+      get(contact, ['cozyMetadata', 'sync', contactAccountId, 'id']) ===
+      googleContact.resourceName
+  )
+}
+
+const updateCozyMetadata = (
+  cozyContact,
+  etag,
+  resourceName,
+  contactAccountId
+) => {
+  const now = new Date().toISOString()
+  const updatedByApps = union(
+    get(cozyContact, 'cozyMetadata.updatedByApps', []),
+    [APP_NAME]
+  )
+
+  return {
+    ...cozyContact,
+    cozyMetadata: {
+      ...cozyContact.cozyMetadata,
+      updatedAt: now,
+      updatedByApps,
+      sync: {
+        [contactAccountId]: {
+          konnector: APP_NAME,
+          lastSync: now,
+          remoteRev: etag,
+          id: resourceName,
+          contactsAccountsId: contactAccountId
+        }
+      }
+    }
+  }
+}
+
+const shouldCreateOnGoogle = (cozyContact, contactAccountId) => {
+  return (
+    !cozyContact.cozyMetadata.sync ||
+    cozyContact.cozyMetadata.sync[contactAccountId] === undefined
+  )
+}
+
+const shouldCreateOnCozy = (cozyContact, googleContact) =>
+  !cozyContact && !googleContact.metadata.deleted
+
+const hasRevChanged = (mergedContact, cozyContact, contactAccountId) => {
+  const cozyRev = get(cozyContact, [
+    'cozyMetadata',
+    'sync',
+    contactAccountId,
+    'remoteRev'
+  ])
+  const mergedRev = get(mergedContact, [
+    'cozyMetadata',
+    'sync',
+    contactAccountId,
+    'remoteRev'
+  ])
+  return mergedRev !== cozyRev
+}
+
+const synchronizeContacts = async (
+  contactAccountId,
+  cozyContacts,
+  googleContacts,
+  cozyUtils,
+  googleUtils
 ) => {
-  const {
-    afterSave,
-    findRemoteDocument,
-    save,
-    shouldSave
-  } = synchronizationStrategy
+  const result = {
+    google: {
+      created: 0,
+      updated: 0,
+      deleted: 0
+    },
+    cozy: {
+      created: 0,
+      updated: 0,
+      deleted: 0
+    }
+  }
   try {
-    return Promise.all(
-      localContacts.map(async localContact => {
-        const remoteContact = findRemoteDocument(localContact, remoteContacts)
-        if (shouldSave(localContact, remoteContact)) {
-          const resp = await save(localContact)
-          const createdId = await afterSave(localContact, resp.data)
-          return {
-            created: true,
-            id: createdId
-          }
+    await Promise.all(
+      cozyContacts.map(async cozyContact => {
+        const googleContact = await findGoogleContact(
+          cozyContact,
+          googleContacts,
+          contactAccountId
+        ) // TODO also query the remote if not in googleContacts
```

Comment:
Why? To check if the contact was not created during the process ?

Reply:
No, it's for the case where a contact has changed cozy-side but not google-side. In that case the contact isn't in `googleContacts` and we can't merge them.

But as I write this, I realize that it's probably not a problem, we can simply use the cozy information to update google's side. We should add a test to confirm that but… we might not need this after all \o/